### PR TITLE
Register macro expansion capabilities

### PR DESCRIPTION
**Pull request by maciektr**
commit-id:3dc2a19c

---

**Stack**:
- #1200
- #1195 ⬅

⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

### CODE REVIEW

File: plugins/cairo-lang-macro/src/lib.rs [hunk 1]
```diff
@@ -3,13 +3,93 @@ pub use cairo_lang_macro_attributes::*;
...
+    pub fun: ExpansionFunc,
+}
+
+type ExpansionFunc = fn(TokenStream) -> ProcMacroResult;
+
+/// Distributed slice for storing procedural macro code expansion capabilities.
+///
+/// Each element denotes name of the macro, and the expand function pointer.
+#[distributed_slice]
+pub static MACRO_DEFINITIONS_SLICE: [ExpansionDefinition];
```
Comment (mkaput) [inline plugins/cairo-lang-macro/src/lib.rs:29]: shouldn't this be doc(hidden)?

File: plugins/cairo-lang-macro/src/lib.rs [hunk 2]
```diff
@@ -3,13 +3,93 @@ pub use cairo_lang_macro_attributes::*;
...
+///
+/// # Safety
+#[no_mangle]
+pub unsafe extern "C" fn list_expansions() -> StableExpansionsList {
+    let list = MACRO_DEFINITIONS_SLICE
+        .iter()
+        .map(|m| m.clone().into_stable())
+        .collect();
+    StableSlice::new(list)
+}
```
Comment (mkaput) [inline plugins/cairo-lang-macro/src/lib.rs:51]: doc(hidden)?

File: plugins/cairo-lang-macro/src/lib.rs [hunk 3]
```diff
@@ -3,13 +3,93 @@ pub use cairo_lang_macro_attributes::*;
...
+/// of the dynamic library re-exporting it.
+///
+/// # Safety
+#[no_mangle]
+pub unsafe extern "C" fn free_expansions_list(list: StableExpansionsList) {
+    let v = list.into_owned();
+    v.into_iter().for_each(|v| {
+        ExpansionDefinition::free_owned(v);
+    });
+}
```
Comment (mkaput) [inline plugins/cairo-lang-macro/src/lib.rs:65]: doc(hidden)?

File: plugins/cairo-lang-macro/src/lib.rs [hunk 4]
```diff
@@ -3,13 +3,93 @@ pub use cairo_lang_macro_attributes::*;
...
+    let fun = EXPAND_CALLBACK
+        .get(item_name.as_str())
+        .expect("proc macro not found");
+    let result = fun(token_stream);
+    let result: StableProcMacroResult = result.into_stable();
+    cairo_lang_macro_stable::StableResultWrapper {
+        input: stable_token_stream,
+        output: result,
+    }
+}
```
Comment (mkaput) [inline plugins/cairo-lang-macro/src/lib.rs:91]: doc(hidden)?

**Review by mkaput — CHANGES_REQUESTED** [review]

File: plugins/cairo-lang-macro/src/lib.rs [hunk 5]
```diff
@@ -3,13 +3,93 @@ pub use cairo_lang_macro_attributes::*;
...
+/// Each element denotes name of the macro, and the expand function pointer.
+#[distributed_slice]
+pub static MACRO_DEFINITIONS_SLICE: [ExpansionDefinition];
+
+static EXPAND_CALLBACK: Lazy<HashMap<String, ExpansionFunc>> = Lazy::new(|| {
+    MACRO_DEFINITIONS_SLICE
+        .iter()
+        .map(|m| (m.name.to_string(), m.fun))
+        .collect()
+});
```
Comment (mkaput) [inline plugins/cairo-lang-macro/src/lib.rs:36]: Do you really need to invest code into creating this hash map? Why not just find the item linearly in the source vector?

Anyway, I bet that for the amounts of items this collection will get, searching `vec` will be faster than making a hashmap and looking up there

**Review by mkaput — CHANGES_REQUESTED** [review]

File: plugins/cairo-lang-macro-attributes/src/lib.rs [hunk 6]
```diff
@@ -13,19 +13,17 @@ use syn::{parse_macro_input, ItemFn};
 pub fn attribute_macro(_args: TokenStream, input: TokenStream) -> TokenStream {
     let item: ItemFn = parse_macro_input!(input as ItemFn);
     let item_name = &item.sig.ident;
```
[the remark this answers is not in the document]
Comment (maciektr) [reply plugins/cairo-lang-macro-attributes/src/lib.rs:15]: Yes, I thought about it and then forgot, sorry!

File: plugins/cairo-lang-macro/Cargo.toml [hunk 7]
```diff
@@ -16,4 +16,5 @@ repository.workspace = true
 cairo-lang-macro-attributes = { path = "../cairo-lang-macro-attributes" }
 cairo-lang-macro-stable = { path = "../cairo-lang-macro-stable" }
 linkme.workspace = true
+once_cell.workspace = true
```
Comment (mkaput) [inline plugins/cairo-lang-macro/Cargo.toml:19]: unused now

**Review by mkaput — CHANGES_REQUESTED** [review]

**Review by mkaput — APPROVED** [review]

**Pull request merged**